### PR TITLE
[action] add envelopeRLP to  handle RLP-encoded tx

### PR DESCRIPTION
**Pull request by dustinxie**

### CODE REVIEW

File: action/envelope_test.go
Comment (dustinxie) [inline action/envelope_test.go:14]: combine multiple 1-line tests together

File: action/rlp_tx.go
Comment (dustinxie) [inline action/rlp_tx.go:30]: we don't need a new struct{}
these are utility funcs

File: action/sealedenvelope_test.go
Comment (dustinxie) [inline action/sealedenvelope_test.go:22]: combine multiple 1-line tests together

File: action/rlp_tx.go
Comment (CoderZhi) [inline action/rlp_tx.go:34]: return error and leave the handling of the error to where it is called

File: action/rlp_tx.go
Comment (CoderZhi) [inline action/rlp_tx.go:40]: same for this function, return (hash.Hash256, error)

File: action/sealedenvelope.go
Comment (CoderZhi) [inline action/sealedenvelope.go:18]: externChainID is not part of the struct, because it can be told from the encoding for now.

File: action/sealedenvelope.go
Comment (CoderZhi) [inline action/sealedenvelope.go:41]: we should either panic or return the error

File: action/sealedenvelope.go
Comment (CoderZhi) [inline action/sealedenvelope.go:32]: if we keep this function, Envelope.Hash, it will be confusing, because it is just the hash of the protobuf encoding format

File: action/sealedenvelope.go
Comment (CoderZhi) [inline action/sealedenvelope.go:17]: encoding iotextypes.WireEncoding

File: action/sealedenvelope.go
Comment (CoderZhi) [inline action/sealedenvelope.go:94]: if we need to do validation for rlp encoding, should we do it for protobuf as well?

File: config/config.go
Comment (CoderZhi) [inline config/config.go:256]: we may have a more precise name which can better explain itself from its name

File: config/config.go
Comment (CoderZhi) [inline config/config.go:479]: if the way of reading this parameter is so strict, what about _plugins and etc.? Thus, I don't see the benefit.

File: config/config.go
Comment (dustinxie) [reply config/config.go:256]: so do you have a suggestion?

File: config/config.go
Comment (dustinxie) [reply config/config.go:479]: can you give more details? _plugins?
what I did is to make sure the external chain ID is loaded exactly one

File: action/sealedenvelope.go
Comment (dustinxie) [reply action/sealedenvelope.go:94]: for protobuf, validation is done in Envelope.Load(), it will check all possible types.

here is to make sure for RLP case, the struct{} inside must be Transfer or Execution, i.e., to catch someone send in a StakeCreate with encoding = 1

File: action/sealedenvelope.go
Comment (dustinxie) [reply action/sealedenvelope.go:18]: we discussed last night, it cannot be told from the encoding (unless we send in all tx bytes, then you can decode and get it)

for now, we'll just set it according to your network config (4689/4690), and use it to calculate raw hash, and verify against the signature
if signature fail, means this tx is not intended for this network

File: config/config.go
Comment (dustinxie) [reply config/config.go:479]: do you mean like "-plugin=gateway"?

this is doable but I think putting in config.yaml is an easier way for delegates.

File: action/sealedenvelope.go
Comment (dustinxie) [reply action/sealedenvelope.go:41]: Hash() returns 32-byte array

i added a check to catch all-0 hash in Verify()

File: action/sealedenvelope.go
Comment (dustinxie) [reply action/sealedenvelope.go:32]: i think the logic is similar to what we discussed for signed hash?
```
if rlp()
   return RLP's raw hash
return sealed.Envelope.Hash(), which is native protobuf's raw hash
```

File: action/rlp_tx.go
Comment (dustinxie) [reply action/rlp_tx.go:34]: since the caller is sealed.Hash(), it does not return err, so return all-0 hash to indicate error

added a check to catch all-0 hash in Sing() and Verify()

Comment (CoderZhi) [reply action/rlp_tx.go:34]: even this is a private function, it is still a good practice to return the error and let the caller handle the error properly.

File: action/sealedenvelope.go
Comment (CoderZhi) [reply action/sealedenvelope.go:41]: If a sealed envelope could be assembled, we shouldn't expect any error here. Thus, we should panic, instead of having some hidden knowledge like ZeroHash

File: action/sealedenvelope.go
Comment (CoderZhi) [reply action/sealedenvelope.go:98]: ok

File: action/action.go
Comment (CoderZhi) [inline action/action.go:44]: how could envelope.hash be zero hash?

File: action/sealedenvelope.go
Comment (CoderZhi) [reply action/sealedenvelope.go:18]: besides https://github.com/iotexproject/iotex-core/pull/2592/files#diff-fda8afdb7165ec30b8d2c40dfd1963def0ffeb117a3d9098167b8139c72527caR97, where else we set the externChainID? If we can use confg.ExternChainID() there for rlp encoding, why cannot we do it in the struct?

File: action/sealedenvelope.go
Comment (CoderZhi) [reply action/sealedenvelope.go:32]: Envelope.Hash is only used in Sign and Verify, in other word, Hash doesn't need to be an essential public API of Envelope.
Same for https://github.com/iotexproject/iotex-core/pull/2592/files#diff-fda8afdb7165ec30b8d2c40dfd1963def0ffeb117a3d9098167b8139c72527caR25, RawHash doesn't expose to external packages

File: config/config.go
Comment (CoderZhi) [reply config/config.go:256]: https://chainid.network/ EVM_NETWORK_ID

File: config/config.go
Comment (CoderZhi) [reply config/config.go:479]: think about all the scenarios how this function will be called (we don't need this function, but set the _evmNetworkID in the New function), these complicated handlings are redundant.

File: config/config.go
Comment (CoderZhi) [inline config/config.go:535]: I can see why we have this function for now. But ideally, we should have a Signer and a Verifier which were created with cfg.EVMNetworkID. Global function is not a good choice, but we can leave it to future.

File: action/action.go
Comment (dustinxie) [reply action/action.go:44]: ok

File: action/sealedenvelope.go
Comment (dustinxie) [reply action/sealedenvelope.go:18]: that's the only place we set externChainID

what do you mean exactly "why cannot we do it in the struct?", as said above, that is the ONLY place where we need to set externChainID, since it only applies for IsRLP() case

File: action/sealedenvelope.go
Comment (dustinxie) [reply action/sealedenvelope.go:32]: ok, right, RawHash() need not expose

File: action/sealedenvelope.go
Comment (dustinxie) [reply action/sealedenvelope.go:41]: yes, ideally, if the signature is correct, rlpSignedHash() will not err.

i feel fallback to 0-hash is safer, b/c 0-hash will fail Verify(), and program will catch the error, and continue. Panic could bring down a node in case a tx somehow caused error in rlpSignedHash() -- though in theory it should not happen

File: config/config.go
Comment (dustinxie) [reply config/config.go:256]: i don't want to explicitly call out the name EVM, also this extern chain ID can in theory be used for another type of wire encoding format -- though that's not very likely in the near future

File: config/config.go
Comment (dustinxie) [reply config/config.go:479]: again my purpose is to make sure the external chain ID is loaded exactly one, i think it's better/safer to do that

File: action/sealedenvelope.go
Comment (dustinxie) [reply action/sealedenvelope.go:32]: will have another PR to remove Envelope.Hash

File: action/sealedenvelope.go
Comment (CoderZhi) [inline action/sealedenvelope.go:68]: returntype iotextypes.WireEncoding

File: action/sealedenvelope.go
Comment (CoderZhi) [inline action/sealedenvelope.go:45]: we should panic here, but let's leave it to next PR

File: action/sealedenvelope.go
Comment (CoderZhi) [inline action/sealedenvelope.go:78]: redundant type casting?

File: config/config.go
Comment (CoderZhi) [reply config/config.go:479]: in what case this struct or function call could be thread unsafe? I don't see it

File: config/config.go
Comment (CoderZhi) [reply config/config.go:256]: it should be explicit and shouldn't be shared by other encoding type

File: action/sealedenvelope.go
Comment (CoderZhi) [inline action/sealedenvelope.go:31]: this is the only place we need to call config.ExternChainID()

**Review by CoderZhi — APPROVED** [review]
Unblock this PR, but the problems in the comments should be fixed addressed in next PR.

File: action/sealedenvelope.go
Comment (dustinxie) [reply action/sealedenvelope.go:31]: also rlpSignedHash()

File: action/sealedenvelope.go
Comment (dustinxie) [reply action/sealedenvelope.go:78]: fixed

File: action/sealedenvelope.go
Comment (dustinxie) [reply action/sealedenvelope.go:45]: node receive tx thru API endpoint --> tx.LoadProto() --> actpool.Add(tx) --> call Verify(tx), if pass, tx is added to actpool, if fail, tx is rejected

if panic here, Verify() will panic, and node stop running
return all-0 hash will make Verify() fail gracefully, reject this tx, and continue running. i think this is better than panic

File: action/sealedenvelope.go
Comment (dustinxie) [reply action/sealedenvelope.go:31]: by saving externChainID in LoadProto() once, all later func calls like `rawHash(), Hash()` become self-contained so they don't need to call outside `config.ExternChainID()` again

**Pull request merged**
